Add draft of lsif-go blog post (#1410)
sourcegraph/about: blogposts/2020/3.19/lsif-go.md

---
title: "lsif-go improvements"
author: Eric Fritz
publishDate: 2020-08-17T08:00-07:00
tags: [blog]
slug: todo
heroImage: /blog/3.19-todo.png
published: true
---

The release of Sourcegraph 3.19 happens to coincide with the [one-point-oh release](https://github.com/sourcegraph/lsif-go/releases/tag/v1.0.0) of [lsif-go](https://github.com/sourcegraph/lsif-go). This release boasts impressive speed improvements that brings intelligence to your Go code faster than ever before.

> The code intelligence is hard at work trying to bring the _time to intelligence_ as close to zero as possible. We want your code to always have up-to-date precise code intelligence on the newest commits to help aid users browsing unfamiliar code and code reviewers trying to understand new code in pull requests. Indexing speed is a _huge_ part of this, and we've made a great leap here.

Reviewer: Asterisk and add a definition somewhere, when I first heard this term used in our meeting I didn't know exactly what you meant.

Author: Can you take another look here?

As an illustrative example, checkout the visual difference between indexing our [main repository](https://github.com/sourcegraph/sourcegraph) with lsif-go v0.9 and indexing it with lsif-go v1.0.

> #### Old and busted (v0.9.0)

Reviewer: I suggest to use a different adjective here then "busted" - busted implies broken and I don't think we want to indicate that lsif-go was broken, just much slower. I suggest something like "sluggish" or even better to appear less "old" friendly - to match the title "the new hotness" with a similar paraphrase "the slow dinosaur" or something like that.

Reviewer 2: for reference: https://www.youtube.com/watch?v=ha-uagjJQ9k

might still be worthwhile to change though :stuck_out_tongue:

Reviewer: You are way cooler then me Garo, I'm just all textbook nerd 📖

> ![lsif-go v0.9](https://storage.googleapis.com/sourcegraph-assets/blog/3.19/lsif-go-ss-0.9.gif)

Reviewer 2: not sure how easy this would be, but could you make these adjacent in one gif? i had some network loading weirdness so that the old and busted finished first

Author: LOL oh no. Yeah I can try to do that.

Author: Replaced the gifs with asciicasts.

#### The new hotness (v1.0.0)

![lsif-go v1.0](https://storage.googleapis.com/sourcegraph-assets/blog/3.19/lsif-go-ss-1.0.gif)

# Performance comparison

Unfortunately, indexing our own code isn't so worthy of celebration. It's a medium-sized repository home to 258k lines of Go code. The old busted tool did a fine job at this small scale, even with its many inefficiencies.

![a repo for ants?](https://i.imgflip.com/4bs5q7.jpg)

What we need to do is go **big**.

> Our goal is to _efficiently_ bring code intelligence to the massive engineering teams with a mind-boggling amount of code and a breathtaking frequency of updates to that code. Our goal is to help humans effectively navigate code bases with a constantly changing topology so they can be sure of their footing and concentrate on the task at hand.

Reviewer 2: for conciseness
```suggestion
We want to help massive engineering teams effectively navigate huge code bases with a constantly changing topology so they can be sure of their footing and concentrate on the task at hand.
```

Author: It wasn't written for conciseness but I'll take the suggestion if it's seconded.

Reviewer: I think the way it's currently written is fine.

So how do we perform at scale?

<div class="no-shadow">
  <img src="https://storage.googleapis.com/sourcegraph-assets/blog/3.19/lsif-go-perf-3.png" alt="performance comparison">
</div>

Much better now, thanks for asking.

In v0.9, it took around 7 1/2 minutes to index the [Go AWS SDK](https://sourcegraph.com/github.com/aws/aws-sdk-go); in v1.0, it takes 24 seconds. In v0.9, it took nearly 7 hours to index 56 million lines of code; in v1.0, it takes under 20 minutes.

**The indexer takes only 5% of the time it used to.**

Indexing speed is _so_ important for code bases undergoing constant change. Stale, hours-old code intelligence on a monorepo at scale is about as useful as using a map of Pangea to find your way home. This is why the code intelligence team is dedicated to [increasing the efficiency of every part of the stack](https://about.sourcegraph.com/blog/optimizing-a-code-intel-backend) to make sure your maps are always accurate.

> <style>

Reviewer: When I am in preview mode this still displays as HTML - is that intended?

Author: You have to run the site locally for it to render properly.
